lib/chef/provider/aws_eip_address.rb

```diff
@@ -25,14 +25,14 @@ def create_aws_object
           Chef::Log.debug "Since associate_to_vpc is not specified and instance #{new_resource.machine} (#{desired_instance.id}) and #{associate_to_vpc ? "is" : "is not"} in a VPC, setting associate_to_vpc to #{associate_to_vpc}."
         end
       end
-      new_resource.driver.ec2.elastic_ips.create vpc: new_resource.associate_to_vpc
+      new_resource.driver.ec2.allocate_address vpc: new_resource.associate_to_vpc
     end
   end
 
   def update_aws_object(elastic_ip)
     if !new_resource.associate_to_vpc.nil?
-      if !!new_resource.associate_to_vpc != !!elastic_ip.vpc?
-        raise "#{new_resource.to_s}.associate_to_vpc = #{new_resource.associate_to_vpc}, but actual IP address has vpc? set to #{elastic_ip.vpc?}.  Cannot be modified!"
+      if !!new_resource.associate_to_vpc != !!(elastic_ip.domain == 'vpc')
```

Comment:
I think we can clean this up a bit. `==` will always return a TrueClass or FalseClass, which allows us to simplify to:

```
if !!new_resource.associate_to_vpc != (elastic_ip.domain == 'vpc')
```

Also, since associate_to_vpc is defined in the resource as:

```
 attribute :associate_to_vpc, kind_of: [TrueClass, FalseClass]
```

I believe Chef will ensure it is one of those two (and we have a nil check at line 33 so that case is also covered).

Thus, we can further simplify to:

```
if new_resource.associate_to_vpc != (elastic_ip.domain == 'vpc')
```